src/sidebar/services/autosave.js

```diff
@@ -0,0 +1,62 @@
+/**
+ * A service for automatically saving new highlights.
+ */
+import { retryPromiseOperation } from '../util/retry';
+
+// @ngInject
+export default function autosaveService(annotationsService, store) {
+  // A Set of annotation $tags that have save requests in-flight
+  const saving = new Set();
+
+  // A Set of annotation $tags that have failed to save after retries
+  const failed = new Set();
+
+  /**
+   * Determine whether we should try to send a save request for the highlight
+   * indicated by `htag`
+   *
+   * @param {string} htag - The local unique identifier for the unsaved highlight
+   * @return {boolean}
+   */
+  const shouldSaveHighlight = htag => {
+    return !saving.has(htag) && !failed.has(htag);
+  };
+
+  /**
+   * Store-subscribed call back. Automatically save new highlights.
+   */
+  const autosaveNewHighlights = () => {
+    const newHighlights = store.newHighlights() || [];
```

Comment:
Won't `store.newHighlights()` return an empty array anyway if there are no new highlights, making the `|| []` redundant?